source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,282 @@
+    const GzipV2CompressionStrategyEnum& compression_strategy) {
+  switch (compression_strategy) {
+  case GzipV2CompressionStrategyEnum::Gzip_CompressionStrategy_RLE:
+    return ZlibCompressionStrategyEnum::Rle;
+  case GzipV2CompressionStrategyEnum::Gzip_CompressionStrategy_FILTERED:
+    return ZlibCompressionStrategyEnum::Filtered;
+  case GzipV2CompressionStrategyEnum::Gzip_CompressionStrategy_HUFFMAN:
+    return ZlibCompressionStrategyEnum::Huffman;
+  default:
+    return ZlibCompressionStrategyEnum::Standard;
+  }
+}
+
+std::unordered_set<std::string>
+GzipFilterConfig::contentTypeSet(const Protobuf::RepeatedPtrField<std::string>& types) {
+  return types.empty() ? defaultContentEncoding()
+                       : std::unordered_set<std::string>(types.cbegin(), types.cend());
+}
+
+uint64_t GzipFilterConfig::contentLengthUint(Protobuf::uint32 length) {
+  return length >= MinimumContentLength ? length : MinimumContentLength;
+}
+
+uint64_t GzipFilterConfig::memoryLevelUint(Protobuf::uint32 level) {
+  return level > 0 ? level : DefaultMemoryLevel;
+}
+
+uint64_t GzipFilterConfig::windowBitsUint(Protobuf::uint32 window_bits) {
+  return (window_bits > 0 ? window_bits : DefaultWindowBits) | GzipHeaderValue;
+}
+
+GzipFilter::GzipFilter(const GzipFilterConfigSharedPtr& config)
+    : skip_compression_{true}, compressed_data_(), compressor_(), config_(config) {}
+
+FilterHeadersStatus GzipFilter::decodeHeaders(HeaderMap& headers, bool) {
+  skip_compression_ = !isAcceptEncodingAllowed(headers);
+  return FilterHeadersStatus::Continue;
+}
+
+FilterHeadersStatus GzipFilter::encodeHeaders(HeaderMap& headers, bool end_stream) {
+  if (!end_stream && !skip_compression_ && isMinimumContentLength(headers) &&
+      isContentTypeAllowed(headers) && isCacheControlAllowed(headers) && isEtagAllowed(headers) &&
+      isLastModifiedAllowed(headers) && isTransferEncodingAllowed(headers) &&
+      !headers.ContentEncoding()) {
+
+    insertVary(headers);
+    headers.removeContentLength();
+    headers.insertContentEncoding().value(Http::Headers::get().ContentEncodingValues.Gzip);
+    compressor_.init(config_->compressionLevel(), config_->compressionStrategy(),
+                     config_->windowBits(), config_->memoryLevel());
+
+  } else {
+    skip_compression_ = true;
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+// TODO(gsagula): We need to test whether or not the compressed output is larger than
+// the uncompressed input. This could potentially be done by forcing flush on the first
+// few buffers received and keeping a copy of the uncompressed ones. If compression is
+// below certain expected ratio, the compressed buffer is disposed and the original
+// response/headers are passed to the client instead.
+FilterDataStatus GzipFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (skip_compression_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  const uint64_t n_data{data.length()};
```

Comment:
I thought braced initialization is always preferred over assignment, because it allows exceptions to be raised at compile time if any.

http://en.cppreference.com/w/c/language/scalar_initialization

Reply:
Ok :)